Clean up directory structure

gnunicorn opened the pull request:
When working [on getting our rust work represented on the website](https://github.com/libp2p/website/pull/79), I noticed that our repo has so many crates now, that it is a bit overwhelming and confusing to look at (especially if you are new to the project). Inspired by the way libp2p generally divides them into bundles, I propose moving all crates but core and the meta-crate into subdirectories according to their area (inspired by the bundles) and removing `circular-buffer` as it is not needed anymore.

Leaving us with a much cleaner directory structure:
```
 - core
 - libp2p
 - muxers
 - protocols
 - transports
 - utils
```

Secondly, I request feedback on the following decisions within that proposal:
 - unpublished crates `core` directly depends on, can be found within the `core/` directory
 - `dns` is in `protocols` at the moment, but I feel weird about it, as it isn't any protocol run on top of our transports, though `kad` is a distinct protocol that runs on top of transports, similarly to `dns` its job is more about discovery, so I am wondering whether these should have their own first-level category instead
 - I've also put `secio` into `protocols` as it, to my understanding, builds on top of and requires a transport to function and I am not aware it does any muxing by itself (which would be different for `quic`)

gnunicorn commented:
> Other than that, I don't think we need a `transports/modifiers` directory. Putting everything in `transports` is good enough I'd say.

Yeah, I thought about that, too and considering there are only 2, I've moved them up. However, I feel they are still distinct and - should we come up with more - might want to reconsider splitting them into `raw` connections and those that can only do stuff, if they have an underlying transport.

(change made)

----

This is also were the other things gets hairy:

> For me the straight-forward definition of "transport" would be something that handles a component of a multiaddress: TCP, IP, UDP, QUIC, DNS, p2p-circuit, etc.

I can very much follow train of thought, but it implies, that our definition of transport and that of mutliaddr match. Which, just looking at [their public example](https://multiformats.io/multiaddr/), doesn't really match our understanding of it - http really is more of a specific protocol than a transport, no?

```
/dns4/foo.com/tcp/80/http/bar/baz.jpg
```

I'm currently thinking of going more towards the understanding of what _establishes_ a channel and passes information through that channel (`transport`) versus what _uses_ an established channel for other purposes. Under this scheme, `kad`, `relay` and `dns` are all to be considered `transports`,  even though `relay` might _use_ a transport to do so, its primary job is still to _establish the channel_ and pass information through it.

I still think we might want to consider `kad` and `dns` some certain `transport` form (like `resolvers`), as they help with establishing but are not needed to _passing information through_ after.

However, I've move all of them into transport for now.

tomaka commented:
> http really is more of a specific protocol than a transport, no?

You can argue that everything is both a transport and a protocol: TCP/IP, WebSockets, etc.

> Under this scheme, kad, relay and dns are all to be considered transports

I don't see how `kad` is a transport under this scheme. `kad` only allows us to retrieve information about the layout of the network and fill a peer store with it. If you're thinking about translating /p2p/ addresses into multiaddresses, this is done through the peer store.

gnunicorn commented:
> If you're thinking about translating /p2p/ addresses into multiaddresses, this is done through the peer store.

If you don't have a local multiaddr for a given p2p-addr, isn't it resolved through a kad-lookup? Isn't that in the `kad`-crate? If so, I'd consider that resolving and part of establishing a connection, if it isn't involved, then you are right, it shouldn't be.

gnunicorn commented:
Updates:
 - moved `kad` into `protocols`
 - renamed `utils` to `stores`, as it contains only `*store`-crates
 - moved `core/*` into `misc/`, leaving `core` to only `core`-things
 - move `libp2p` to `/`, inspired by `wasm_bindgen`, to have the main repo in the top level, allowing for neat things like `cargo run --example` directly from the repo root; workspace still works as before

tomaka commented:
Why are there conflicts? We haven't merged anything in a long time. :worried: 
Please be extra careful that no source code has been modified in the process.

twittner commented:
The conflicts are due to the merge of https://github.com/libp2p/rust-libp2p/pull/432.
